ktor-network/common/test/io/ktor/network/sockets/tests/UDPSocketTest.kt

```diff
@@ -1,42 +1,29 @@
 /*
-* Copyright 2014-2021 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
-*/
+ * Copyright 2014-2021 JetBrains s.r.o and contributors. Use of this source code is governed by the Apache 2.0 license.
+ */
 
 package io.ktor.network.sockets.tests
 
 import io.ktor.network.selector.*
 import io.ktor.network.sockets.*
+import io.ktor.test.dispatcher.*
+import io.ktor.util.*
 import io.ktor.util.network.*
 import io.ktor.utils.io.core.*
+import kotlinx.atomicfu.*
 import kotlinx.coroutines.*
-import kotlinx.coroutines.debug.junit4.*
-import org.junit.*
-import java.net.*
-import kotlin.coroutines.*
-import kotlin.io.use
 import kotlin.test.*
-import kotlin.test.Test
 
-class UDPSocketTest : CoroutineScope {
-    private val testJob = Job()
-    private val selector = ActorSelectorManager(Dispatchers.Default + testJob)
+// TODO: some tests are logging the following exception (tests still green). Not sure why this is happening, this was not the case before rebasing.
+//  io.ktor.utils.io.errors.PosixException.BadFileDescriptorException: EBADF (9): Bad file descriptor
```

Comment:
Not sure why these errors are logged. This was not the cause in an older version of this PR. Would appreciate some help with this.

Reply:
It seems that these errors are only logged when the socket is closed. I think it is caused by https://github.com/ktorio/ktor/commit/ca07f913904b79e4c535ed44ac3aa2f47ec4e74d, but I am not sure.

@cy6erGn0m could you help here, please? Do you know why a `BadFileDescriptorException` is logged? See complete stack trace below:

```
Uncaught Kotlin exception: kotlinx.coroutines.CoroutinesInternalError: Fatal exception in coroutines machinery for DispatchedContinuation[WorkerCoroutineDispatcherImpl@2cb47428, Continuation @ $start$lambda-0COROUTINE$16]. Please read KDoc to 'handleFatalException' method and report this incident to maintainers
    at 0   test.kexe                           0x0000000102181f1e kfun:kotlin.Error#<init>(kotlin.String?;kotlin.Throwable?){} + 126 (/Users/teamcity1/teamcity_work/290aee0e088a1666/runtime/src/main/kotlin/kotlin/Exceptions.kt:14:63)
    at 1   test.kexe                           0x000000010237a8de kfun:kotlinx.coroutines.CoroutinesInternalError#<init>(kotlin.String;kotlin.Throwable){} + 126 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/Exceptions.common.kt:28:77)
    at 2   test.kexe                           0x00000001023ccc77 kfun:kotlinx.coroutines.DispatchedTask#handleFatalException(kotlin.Throwable?;kotlin.Throwable?){} + 887 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/internal/DispatchedTask.kt:144:22)
    at 3   test.kexe                           0x00000001023cc89c kfun:kotlinx.coroutines.DispatchedTask#run(){} + 3484 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/internal/DispatchedTask.kt:115:13)
    at 4   test.kexe                           0x0000000102375dc0 kfun:kotlinx.coroutines.EventLoopImplBase#processNextEvent(){}kotlin.Long + 896 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/EventLoop.common.kt:277:22)
    at 5   test.kexe                           0x00000001023f012f kfun:kotlinx.coroutines#runEventLoop(kotlinx.coroutines.EventLoop?;kotlin.Function0<kotlin.Boolean>){} + 831 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/native/src/Builders.kt:80:40)
    at 6   test.kexe                           0x00000001023f9e39 kfun:kotlinx.coroutines.WorkerCoroutineDispatcherImpl.start$lambda-0#internal + 409 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/native/src/Workers.kt:49:17)
    at 7   test.kexe                           0x00000001023f9f80 kfun:kotlinx.coroutines.WorkerCoroutineDispatcherImpl.$start$lambda-0$FUNCTION_REFERENCE$35.invoke#internal + 64 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/native/src/Workers.kt:47:24)
    at 8   test.kexe                           0x00000001023fa090 kfun:kotlinx.coroutines.WorkerCoroutineDispatcherImpl.$start$lambda-0$FUNCTION_REFERENCE$35.$<bridge-UNN>invoke(){}#internal + 64 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/native/src/Workers.kt:47:24)
    at 9   test.kexe                           0x00000001021b8299 WorkerLaunchpad + 185 (/Users/teamcity1/teamcity_work/290aee0e088a1666/runtime/src/main/kotlin/kotlin/native/concurrent/Internal.kt:70:54)
    at 10  test.kexe                           0x000000010234299a _ZN6Worker19processQueueElementEb + 3162
    at 11  test.kexe                           0x0000000102341d26 _ZN12_GLOBAL__N_113workerRoutineEPv + 54
    at 12  libsystem_pthread.dylib             0x00007fff203f2954 _pthread_start + 224
    at 13  libsystem_pthread.dylib             0x00007fff203ee4a7 thread_start + 15
Caused by: io.ktor.utils.io.errors.PosixException.BadFileDescriptorException: EBADF (9): Bad file descriptor
    at 0   test.kexe                           0x0000000102188cff kfun:kotlin.Throwable#<init>(kotlin.String?){} + 95 (/Users/teamcity1/teamcity_work/290aee0e088a1666/runtime/src/main/kotlin/kotlin/Throwable.kt:23:37)
    at 1   test.kexe                           0x000000010218204d kfun:kotlin.Exception#<init>(kotlin.String?){} + 93 (/Users/teamcity1/teamcity_work/290aee0e088a1666/runtime/src/main/kotlin/kotlin/Exceptions.kt:23:44)
    at 2   test.kexe                           0x00000001020dfec9 kfun:io.ktor.utils.io.errors.PosixException#<init>(kotlin.Int;kotlin.String){} + 105 (/Users/thomas/GitHub/ktor/ktor-io/posix/src/io/ktor/utils/io/errors/PosixErrors.kt:40:78)
    at 3   test.kexe                           0x00000001020dff72 kfun:io.ktor.utils.io.errors.PosixException.BadFileDescriptorException#<init>(kotlin.String){} + 98 (/Users/thomas/GitHub/ktor/ktor-io/posix/src/io/ktor/utils/io/errors/PosixErrors.kt:42:64)
    at 4   test.kexe                           0x00000001020e1116 kfun:io.ktor.utils.io.errors.PosixException.Companion#forErrno(kotlin.Int;kotlin.String?){}io.ktor.utils.io.errors.PosixException + 1910 (/Users/thomas/GitHub/ktor/ktor-io/posix/src/io/ktor/utils/io/errors/PosixErrors.kt:115:26)
    at 5   test.kexe                           0x00000001020e1bc9 kfun:io.ktor.utils.io.errors.PosixException.Companion#forErrno$default(kotlin.Int;kotlin.String?;kotlin.Int){}io.ktor.utils.io.errors.PosixException + 409 (/Users/thomas/GitHub/ktor/ktor-io/posix/src/io/ktor/utils/io/errors/PosixErrors.kt:99:16)
    at 6   test.kexe                           0x00000001020f30d8 kfun:io.ktor.network.selector.SelectorHelper.selectionLoop#internal + 2344 (/Users/thomas/GitHub/ktor/ktor-network/posix/src/io/ktor/network/selector/SelectUtils.kt:74:13)
    at 7   test.kexe                           0x00000001020f512e kfun:io.ktor.network.selector.SelectorHelper.$start$lambda-0COROUTINE$16.invokeSuspend#internal + 318 (/Users/thomas/GitHub/ktor/ktor-network/posix/src/io/ktor/network/selector/SelectUtils.kt:44:13)
    at 8   test.kexe                           0x00000001021aa144 kfun:kotlin.coroutines.native.internal.BaseContinuationImpl#resumeWith(kotlin.Result<kotlin.Any?>){} + 788 (/Users/teamcity1/teamcity_work/290aee0e088a1666/runtime/src/main/kotlin/kotlin/coroutines/ContinuationImpl.kt:30:39)
    at 9   test.kexe                           0x00000001023cc637 kfun:kotlinx.coroutines.DispatchedTask#run(){} + 2871 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/internal/DispatchedTask.kt:106:71)
    at 10  test.kexe                           0x0000000102375dc0 kfun:kotlinx.coroutines.EventLoopImplBase#processNextEvent(){}kotlin.Long + 896 (/opt/buildAgent/work/44ec6e850d5c63f0/kotlinx-coroutines-core/common/src/EventLoop.common.kt:277:22)
    ... and 9 more common stack frames skipped

```

Reply:
It looks like the select is executed on the disposed descriptor. Could you verify that descriptor is still valid and didn't disposed concurrently?

Reply:
Yes looks like that is the case because if I do not close the descriptor in the test then there are no errors logged. I am trying to figure out how the select code works but it is quite difficult to understand for me.

Reply:
I tried to fix it (see: https://github.com/Thomas-Vos/ktor/commit/a01e06b620eabdef7267c77838e54c53371bac49) but unfortunately the error is still sometimes logged. I would appreciate some help with this.

Reply:
Another attempt to fix this which looks like it is working: https://github.com/Thomas-Vos/ktor/commit/7ba2034b6696465bd16a86bf59e8c39e013e5599

Reply:
Pushed the fix to this PR, would be great if you could check it out @e5l.